File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/UpdateCategoryOrderController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.dto.CategoryDTO
import net.animeta.backend.model.User
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.CategorySerializer
import net.animeta.backend.service.CategoryMutations
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class UpdateCategoryOrderController(
    private val categoryMutations: CategoryMutations,
    private val categorySerializer: CategorySerializer
) {
    data class Params(
        val categoryIds: List<Int>
    )
    data class Result(
        val categories: List<CategoryDTO>
    )

    @PostMapping("/v3/UpdateCategoryOrder")
    fun handle(@RequestBody params: Params, @CurrentUser currentUser: User): Result {
        val updatedCategories = categoryMutations.updatePositions(currentUser, params.categoryIds)
        return Result(updatedCategories.map { categorySerializer.serialize(it) })
    }
}

File: backend-ts/src/controllers/update_category_order.controller.ts
import { Body, Controller, HttpCode, Post } from "@nestjs/common";
import { CategoryDTO } from 'shared/types';
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { CategorySerializer } from "src/serializers/category.serializer";
import { CategoryService } from "src/services/category.service";
import { Connection } from "typeorm";

type Params = {
  categoryIds: number[];
}

type Result = {
  categories: CategoryDTO[];
}

@Controller('/api/v4/UpdateCategoryOrder')
export class UpdateCategoryOrderController {
  constructor(
    private connection: Connection,
    private categoryService: CategoryService,
    private categorySerializer: CategorySerializer,
  ) {}

  @Post()
  @HttpCode(200)
  async handle(
    @Body() params: Params,
    @CurrentUser() currentUser: User,
  ): Promise<Result> {
    return await this.connection.transaction(async em => {
      const categories = await this.categoryService.updateCategoryOrder(em, currentUser, params.categoryIds)
      return { categories: categories.map(it => this.categorySerializer.serialize(it)) }
    })
  }
}
